File: src/main/java/tech/macil/minecraft/rcon/util/NLRequiringBufferedReader.java
package tech.macil.minecraft.rcon.util;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.Reader;

// Just like BufferedReader but only reads lines that ends in '\n'.
public class NLRequiringBufferedReader extends BufferedReader {
    public NLRequiringBufferedReader(Reader in) {
        super(in);
    }

    public NLRequiringBufferedReader(Reader in, int sz) {
        super(in, sz);
    }

    public String readLine() throws IOException {
        StringBuilder sb = new StringBuilder();
        while (true) {
            int r = read();
            if (r == -1) {
                return null;
            }
            if (r == '\n') {
                return sb.toString();
            }
            sb.append((char) r);
        }
    }
}

File: src/main/kotlin/tech/macil/minecraft/rcon/util/NLRequiringBufferedReader.kt
package tech.macil.minecraft.rcon.util

import java.io.BufferedReader
import java.io.IOException
import java.io.Reader

// Just like BufferedReader but only reads lines that ends in '\n'.
class NLRequiringBufferedReader : BufferedReader {
    constructor(r: Reader) : super(r)

    constructor(r: Reader, sz: Int) : super(r, sz)

    override fun readLine(): String? {
        val sb = StringBuilder()
        while (true) {
            val r = read()
            if (r == -1) {
                return null
            }
            if (r == '\n'.toInt()) {
                return sb.toString()
            }
            sb.append(r.toChar())
        }
    }
}
